Deploying 3.0.1-20180803 on wildfly 13: java.lang.IllegalArgumentException: URI scheme is not \"file\"

Deploying rundeck 3.0.1-20180803 on wildfly 13 (ubuntu, java 8) ,  results in this error:

{"WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:" => {"Operation step-2" => {"WFLYCTL0080: Failed services" => {"jboss.undertow.deployment.default-server.default-host./rundeck" => "java.lang.RuntimeException: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'application': Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: URI scheme is not \"file\"
    Caused by: java.lang.RuntimeException: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'application': Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: URI scheme is not \"file\"
    Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'application': Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: URI scheme is not \"file\"
    Caused by: java.lang.IllegalArgumentException: URI scheme is not \"file\""},"WFLYCTL0288: One or more services were unable to start due to one or more indirect dependencies not being available." => {"Services that were unable to start


Rundeck 2.11 was working fine in wildfly 13.


How to Install:

1. wildfly properties are defined:

rdeck.base=/data/rundeck
RDECK_BASE=/data/rundeck
rundeck.config.location=/data/rundeck/server/config/rundeck-config.properties

(rundeck.config.location was /data/rundeck/etc/rundeck-config.properties with rundeck 2.11)

2. Installing in standalone mode:

The war is in /data/rundeck:
- cd /data/rundeck
- java -jar rundeck.war --installonly


3. Modifying file:
server/config/rundeck-config.properties:

grails.serverURL=https://wildflyURL/rundeck
dataSource.jndiName=java:/jdbc/rundeck

4. Deploying war with jboss-cli.sh.


How to know what URI it is complaining about?

--- a/grails-persistlocale/build.gradle
+++ b/grails-persistlocale/build.gradle
@@ -28,7 +28,7 @@ dependencies {
     compile "org.springframework.boot:spring-boot-autoconfigure"
     compile "org.grails:grails-core"
     compile "org.springframework.boot:spring-boot-starter-actuator"
-    compile "org.springframework.boot:spring-boot-starter-tomcat"
+    provided "org.springframework.boot:spring-boot-starter-jetty"
     compile "org.grails:grails-web-boot"
     compile "org.grails:grails-logging"
     compile "org.grails:grails-plugin-rest"

--- a/rundeckapp/grails-app/init/rundeckapp/init/RundeckInitializer.groovy
+++ b/rundeckapp/grails-app/init/rundeckapp/init/RundeckInitializer.groovy
@@ -66,6 +66,8 @@ class RundeckInitializer {
     File toolsdir;
     File toolslibdir;
 
+    private static boolean vfsDirectoryDetected = false
+
     /**
      * Config properties are defaulted in config-defaults.properties, but can be overridden by system properties
      */
@@ -136,9 +138,13 @@ class RundeckInitializer {
         if(!coreJar.parentFile.exists()) coreJar.parentFile.mkdirs()
         coreJar.createNewFile()
         if(thisJar.isDirectory()) {
-            def coreJarLoc = ((URLClassLoader)Thread.currentThread().contextClassLoader).getURLs().find { it.toString().endsWith(coreJarName) }
-            if(!coreJarLoc && thisJar.name == "classes") {
-                coreJarLoc = new File(thisJar.parentFile,"lib/"+coreJarName)
+            def coreJarLoc = null
+            if(Thread.currentThread().contextClassLoader instanceof URLClassLoader) {
+                coreJarLoc = ((URLClassLoader) Thread.currentThread().contextClassLoader).getURLs().
+                        find { it.toString().endsWith(coreJarName) }
+            }
+            if (!coreJarLoc && thisJar.name == "classes") {
+                coreJarLoc = new File(thisJar.parentFile, "lib/" + coreJarName)
             }
             if(coreJarLoc) {
                 coreJarLoc.withInputStream {
@@ -218,10 +224,13 @@ class RundeckInitializer {
     }
 
     void copyLibsToToolsNonJar(final File toolslibdir, final String[] jarNamesToCopy) {
-        def classlibList = ((URLClassLoader)Thread.currentThread().contextClassLoader).getURLs()
+        def classlibList = []
+        if(Thread.currentThread().contextClassLoader instanceof URLClassLoader) {
+            classlibList = ((URLClassLoader) Thread.currentThread().contextClassLoader).getURLs()
+        }
         jarNamesToCopy.each { jarName ->
             def sourceJar = classlibList.find { it.toString().endsWith(jarName) }
-            if(!sourceJar && thisJar.name == "classes") {  //thisJar is WEB-INF/classes
+            if(thisJar.name == "classes") {  //thisJar is WEB-INF/classes
                 sourceJar = new File(thisJar.parentFile,"lib/"+jarName)
                 if(!sourceJar.exists()) sourceJar = null
             }
@@ -277,7 +286,15 @@ class RundeckInitializer {
             protectionDomain = baseClass.getProtectionDomain()
         }
 
-        final URL location = protectionDomain.getCodeSource().getLocation();
+        URL location = protectionDomain.getCodeSource().getLocation();
+        if(location.toString().startsWith("vfs:")) {
+            //This is probably a jboss based server
+            def vfsFile = location.getContent()
+            vfsDirectoryDetected = true
+            if(location.toString().endsWith("WEB-INF/classes")) {
+                return vfsFile.getPhysicalFile()
+            }
+        }
 
         try {
             return new File(location.toURI());
@@ -320,7 +337,7 @@ class RundeckInitializer {
         String destinationFilePath = props.getProperty(renamedDestFileName+LOCATION_SUFFIX) ?: destDir.absolutePath +"/" + sourceDirPath.relativize(sourceTemplate.parentFile.toPath()).toString()+"/"+renamedDestFileName
 
         File destinationFile = new File(destinationFilePath)
-        if(renamedDestFileName == "log4j.properties" && Environment.isWarDeployed()) {
+        if(renamedDestFileName == "log4j.properties" && Environment.isWarDeployed() && !vfsDirectoryDetected) {
             destinationFile = new File(thisJar.absolutePath,renamedDestFileName)
         }
 
@@ -329,7 +346,7 @@ class RundeckInitializer {
         DEBUG("Writing config file: " + destinationFile.getAbsolutePath());
         expandTemplate(sourceTemplate.newInputStream(),destinationFile.newOutputStream(),props)
 
-        if(renamedDestFileName == "rundeck-config.properties" && Environment.isWarDeployed()) {
+        if(renamedDestFileName == "rundeck-config.properties" && Environment.isWarDeployed() && !vfsDirectoryDetected) {
             List<String> rundeckConfig = destinationFile.readLines()
             destinationFile.withOutputStream { out ->
                 rundeckConfig.each { line ->
@@ -657,7 +674,7 @@ class RundeckInitializer {
      *
      * @param s
      */
-    private void LOG(final String s) {
+    static private void LOG(final String s) {
         System.out.println(s);
     }
     /**
@@ -665,7 +682,7 @@ class RundeckInitializer {
      *
      * @param s
      */
-    private void ERR(final String s) {
+    static private void ERR(final String s) {
         System.err.println("ERROR: " + s);
     }
 

